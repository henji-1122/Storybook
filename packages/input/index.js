// 组件打包入口文件
// 使用组件并将其导出
import wxcInput from './src/input.vue'

// 定义install的方法，后期别人使用时通过Vue.use()方式引入此包
wxcInput.install = Vue => {
  // 全局注册组件
  Vue.component(wxcInput.name, wxcInput) // wxcInput.name:通过组件的名字来获取此组件
}

export default wxcInput