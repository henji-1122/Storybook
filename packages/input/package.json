{
  "name": "wxc-input",
  "version": "0.0.4",
  "description": "",
  "main": "dist/cjs/index.js",
  "module": "dist/es/index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "del": "rimraf dist"
  },
  "keywords": [],
  "author": "WXC",
  "license": "MIT",
  "dependencies": {
    "lodash": "3"
  },
  "gitHead": "06cf8581dfe223b4b06f719cf3f2362a8daef6fb"
}
