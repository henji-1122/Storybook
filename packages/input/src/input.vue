<template>
  <div>
    <input v-bind="$attrs" :type="type" :value="value" @input="handleInput">
  </div>
</template>

<script>
export default {
  name: 'wxcInput',
  inheritAttrs: false,
  props: {
    value: {
      type: String
    },
    type: {
      type: String,
      default: 'text'
    }
  },

  methods: {
    handleInput(event) {
      this.$emit('input', event.target.value)

      // 通知父组件 触发validate方法 进行验证
      // 循环找父组件方法
      const findParent = parent => {
        while(parent){
          // 如果父组件存在，还要判断是不是FormItem组件
          if(parent.$options.name === 'wxcFormItem') { // wxcFormItem:wxcFormItem的name
            break
          }else{
            parent = parent.$parent
          }
        }
        return parent
      }
      // 找到parent后触发父组件的validate事件
      const parent = findParent(this.$parent) // this.$parent:当前组件的父组件是否存在
      if(parent){
        parent.$emit('validate') //触发自定义事件
      }
    }
  }
}
</script>

<style>

</style>