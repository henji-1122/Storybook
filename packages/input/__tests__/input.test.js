// 先导入input组件
import input from '../src/input.vue'
// 引入vue test utils 使用里面的API挂载组件
import { mount } from '@vue/test-utils'
import { TestScheduler } from 'jest'
// Jest不用导入，因为测试文件是被jest加载执行的

// 通过describe创建一个代码块，将input的相关测试都放在这个代码块中
describe('wxc-input', () => { // 代码块中写测试函数
  // 测试input生成的是否是希望的文本框
  test('input-text', () => {
    // mount挂载组件(只是在内存中挂载)
    const wrapper = mount(input)
    // 判断组件生成的html中是否包含type=text
    expect(wrapper.html()).toContain('input type="text"')
  })

  // 测试生成的密码框是否正确
  test('input-password', () => {
    const wrapper = mount(input,{
      propsData: {
        type: 'password'
      }
    })
    expect(wrapper.html()).toContain('input type="password"')
  })

  // 对组件测试时经常要对组件的状态，也就是组件对外公布的数据或方法进行测试
  // 所以下面对value进行测试
  test('input-password', () => {
    const wrapper = mount(input,{
      propsData: {
        type: 'password',
        value: 'admin'
      }
    })
    // 判断生成的组件props对象的value属性值是否是admin
    expect(wrapper.props('value')).toBe('admin')
  })

  // 测试快照，快照是jest中提供的一个方法
  test('input-snapshot', () => {
    const wrapper = mount(input,{
      propsData: {
        type: 'password',
        value: 'admin'
      }
    })
    // 把挂载的组件对应的DOM对象拍一个快照，第一次调用会把expect的值以字符串的方式存储到文本文件中
    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})

