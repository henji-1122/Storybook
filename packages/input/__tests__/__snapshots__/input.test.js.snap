// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wxc-input input-snapshot 1`] = `
<div>
  <input
    type="password"
  />
</div>
`;
